packages/spatial_pooler/src/nupic/research/frameworks/spatial_pooler/spatial_pooler.py

```diff
@@ -0,0 +1,834 @@
+          becoming active. depends on active_duty_cycles via an exponential function.
+
+        - overlaps[i] determines overlap between minicolumn "i" and input vector. overlap = number of connected synapses to input bits which are on.
+
+        - boosted_overlaps[i] = boost_factors[i] * overlaps[i] if spatial pooler is learning, else boosted_overlaps[i] = overlaps[i]
+        '''
+        self.overlap_duty_cycles = np.zeros(self.num_minicolumns, dtype=real_type)
+        self.active_duty_cycles = np.zeros(self.num_minicolumns, dtype=real_type)
+        self.min_overlap_duty_cycles = np.zeros(self.num_minicolumns, dtype=real_type)
+        self.min_percent_overlap_duty_cycles = min_percent_overlap_duty_cycles
+        self.init_connected_percent = 0.5
+        self.boost_strength = boost_strength
+        self.boost_factors = np.ones(self.num_minicolumns, dtype=real_type)
+        self.overlaps = np.zeros(self.num_minicolumns, dtype=real_type)
+        self.boosted_overlaps = np.zeros(self.num_minicolumns, dtype=real_type)
+
+
+        ''' random seed '''
+        if seed == -1:
+            self.seed = np.random.randint(1e10)
+        else:
+            self.seed = seed
+
+        self.generator = np.random.default_rng(seed=self.seed)
+        
+
+        ''' 
+        rows of matrix represent minicolumns: i.e. map from input bits to minicolumns. 
+        columns of matrix represent input bits: i.e. map from minicolumns to input bits.
+
+        - potential_pools[i][j] shows if input bit "j" is in the potential pool of minicolumn "i". a minicolumn can only be connected to inputs in its
+          potential pool.
+
+        - permanences[i][j] shows the permanence for minicolumn "i" to input bit "j".
+
+        - connected_synapses[i][j] shows if minicolumn "i" is connnected to input bit "j".
+
+        - connected_synapse_counts[i] shows the number of connected synapses for minicolumn "i".
+        '''
+        self.potential_pools = np.zeros((self.num_minicolumns, self.num_inputs), dtype=np.bool_) 
+        self.permanences = np.zeros((self.num_minicolumns, self.num_inputs), dtype=real_type)
+        self.connected_synapses = np.zeros((self.num_minicolumns, self.num_inputs), dtype=np.bool_)
+        self.connected_synapses_counts = np.zeros(self.num_minicolumns, dtype=real_type)
+
+
+        for minicolumn_index in range(self.num_minicolumns):
+            potential = self.map_potential(minicolumn_index)
+            self.potential_pools[minicolumn_index, potential.nonzero()[0]] = 1
+            permanence = self.init_permanence(potential)
+            self.update_permanences_for_minicolumn(permanence, minicolumn_index, raise_perm=True)
+
+
+        self.inhibition_radius = 0
+        self.update_inhibition_radius()
+
+    
+    def compute(self, input_vector, learn, active_array):
+        '''
+        primary spatial pooler method. takes an input vector and outputs the indices of the active minicolumns. if learn is True, update the
```

Comment:
Meta question: why do all sentences in comments start with lower case letters instead of the standard upper case?

Reply:
Wanted to keep the "Python" style of coding (probably just something I like to do) -- everything is lowercase, including variable names. No camel case.